src/Microsoft.DotNet.ProjectJsonMigration/Rules/MigrateProjectDependenciesRule.cs

```diff
@@ -10,39 +11,134 @@
         {
             _projectDirectory = migrationSettings.ProjectDirectory;
 
-            var csproj = migrationRuleInputs.OutputMSBuildProject;
-            var projectContext = migrationRuleInputs.DefaultProjectContext;
-            var projectExports = projectContext.CreateExporter("_").GetDependencies(LibraryType.Project);
+            var possibleProjectDependencies = _projectDependencyFinder
+                .FindPossibleProjectDependencies(migrationRuleInputs.DefaultProjectContext.ProjectFile.ProjectFilePath);
 
-            var projectDependencyTransformResults =
-                projectExports.Select(projectExport => ProjectDependencyTransform.Transform(projectExport));
+            var migratedXProjDependencyPaths = MigrateXProjProjectDependencies(migrationSettings, migrationRuleInputs);
+            var migratedXProjDependencyNames = migratedXProjDependencyPaths.Select(p => Path.GetFileNameWithoutExtension(p));
 
-            if (projectDependencyTransformResults.Any())
+
+            AddPropertyTransformsToCommonPropertyGroup(migrationRuleInputs.CommonPropertyGroup);
+            MigrateProjectJsonProjectDependencies(
+                possibleProjectDependencies, 
+                migrationRuleInputs.ProjectContexts, 
+                migratedXProjDependencyNames, 
+                migrationRuleInputs.OutputMSBuildProject);
+        }
+
+        private void ThrowIfUnresolvedDependencies(IEnumerable<ProjectContext> projectContexts, List<ProjectDependency> projectDependencies, IEnumerable<string> migratedXProjDependencyNames)
+        {
+            foreach (var projectContext in projectContexts)
             {
-                AddPropertyTransformsToCommonPropertyGroup(migrationRuleInputs.CommonPropertyGroup);
-                AddProjectDependenciesToNewItemGroup(csproj.AddItemGroup(), projectDependencyTransformResults);
+                var projectExports = projectContext.CreateExporter("_").GetDependencies(LibraryType.Project);
             }
         }
 
-        private void AddProjectDependenciesToNewItemGroup(ProjectItemGroupElement itemGroup, IEnumerable<ProjectItemElement> projectDependencyTransformResults)
+        private IEnumerable<string> MigrateXProjProjectDependencies(MigrationSettings migrationSettings, MigrationRuleInputs migrationRuleInputs)
         {
+            var xproj = migrationRuleInputs.ProjectXproj;
+            if (xproj == null)
+            {
+                MigrationTrace.Instance.WriteLine($"{nameof(MigrateProjectDependenciesRule)}: No xproj file given.");
+                return Enumerable.Empty<string>();
+            }
+
+            var projectReferenceItems = xproj.Items.Where(i => i.ItemType == "ProjectReference");
+            
+            IEnumerable<string> projectReferences = new List<string>();
+            foreach (var projectReferenceItem in projectReferenceItems)
+            {
+                projectReferences = projectReferences.Union(projectReferenceItem.Includes());
```

Comment:
HashSet doesn't seem to provide an AddRange or Union, at least according to VSCode intellisense. I tried that first.

For a small set of strigns like this the perf overhead for constructing the hashset buckets is likely greater than doing n^2 string comparisons, so I'm not worried too much about the perf optimization here.